windowManager: Add wobbly windows.

smspillaz opened the pull request:
Link to libendless_wobbly (from libwobbly) and add a new
EndlessShellFXWobbly class to bridge between a clutter effect
and the wobbly::Model/wobbly::Anchor.

Use endless_shell_fx_wobbly_grab to grab a point on the mesh
and move it around.

Use endless_shell_fx_wobbly_ungrab to release a point on the mesh.

Each of the above functions may only be called once whilst in the
opposing state and cannot be called again whilst in the same state.

EOSShellWobbly is a JavaScript subclass of EndlessShellFXWobbly
which handles integration between the shell itself and the wobbly
effect. It provides API to handle mouse grab / ungrab of a window
and window movement whilst grabbed.

Provide the following new keys in org.gnome.shell:

 'wobbly-spring-k' : Spring K
 'wobbly-spring-friction' : Spring Friction
 'wobbly-slowdown-factor' : Amount to slow effect down
 'wobbly-object-movement-range' : Amount points on the mesh are
  allowed to deviate from their target position once the mesh is
  settled.

https://phabricator.endlessm.com/T17123

cosimoc commented on js/ui/windowManager.js:
```diff
@@ -135,6 +137,76 @@ const DisplayChangeDialog = new Lang.Class({
+    },
+
+    set settings (val)
+    {
```
Move to previous line

cosimoc commented on js/ui/windowManager.js:
```diff
@@ -135,6 +137,76 @@ const DisplayChangeDialog = new Lang.Class({
+
+    set settings (val)
+    {
+         if (val['schema-id'] !== 'org.gnome.shell')
```
I don't see how anything different could happen here; you can remove this code (or just use the global settings here directly)

cosimoc commented on js/ui/windowManager.js:
```diff
@@ -135,6 +137,76 @@ const DisplayChangeDialog = new Lang.Class({
+
+         this._settings = val;
+
+         let binder = Lang.bind (this, function(key, prop) {
```
Remove space before paren (both here and below)

cosimoc commented on js/ui/windowManager.js:
```diff
@@ -135,6 +137,76 @@ const DisplayChangeDialog = new Lang.Class({
+    },
+
+    grabbedByMouse: function() {
+
```
Remove empty line

cosimoc commented on js/ui/windowManager.js:
```diff
@@ -135,6 +137,76 @@ const DisplayChangeDialog = new Lang.Class({
+    ungrabbedByMouse: function() {
+        // Only continue if we have an active grab and change notification
+        // on movement
+        if (this._positionChangedId !== undefined) {
```
Prefer early return

cosimoc commented on js/ui/windowManager.js:
```diff
@@ -2436,4 +2510,53 @@ const WindowManager = new Lang.Class({
     },
+
+    _windowCanWobble: function(window, op) {
+
```
Remove empty line

cosimoc commented on js/ui/windowManager.js:
```diff
@@ -2436,4 +2510,53 @@ const WindowManager = new Lang.Class({
+    },
+
+    _windowGrabbed: function(display, screen, window, op) {
+
```
Remove empty line

cosimoc commented on src/wobbly-effect.cpp:
```diff
@@ -0,0 +1,584 @@
+        delete priv->model;
+
+    if (priv->timeout_id)
+    {
```
I'd love it if we could use an uniform coding style with the rest of `src` for this file.

smspillaz commented:
@cosimoc addressed all comments and rebased

I also tidied up the coding style of wobbly-effect.cpp - from what I understand of the C coding style (I haven't spent much time in it) it should be correct.
